promql/functions.go

```diff
@@ -65,13 +65,121 @@ func funcTime(_ []Vector, _ Matrix, _ parser.Expressions, enh *EvalNodeHelper) (
 	}}, nil
 }
 
+// pickOrInterpolateLeft returns the value at the left boundary of the range.
+// If interpolation is needed (when smoothed is true and the first sample is before the range start),
+// it returns the interpolated value at the left boundary; otherwise, it returns the first sample's value.
+func pickOrInterpolateLeft(floats []FPoint, first int, rangeStart int64, smoothed, isCounter bool) float64 {
+	if smoothed && floats[first].T < rangeStart {
+		return interpolate(floats[first], floats[first+1], rangeStart, isCounter, true)
+	}
+	return floats[first].F
+}
+
+// pickOrInterpolateRight returns the value at the right boundary of the range.
+// If interpolation is needed (when smoothed is true and the last sample is after the range end),
+// it returns the interpolated value at the right boundary; otherwise, it returns the last sample's value.
+func pickOrInterpolateRight(floats []FPoint, last int, rangeEnd int64, smoothed, isCounter bool) float64 {
+	if smoothed && last > 0 && floats[last].T > rangeEnd {
+		return interpolate(floats[last-1], floats[last], rangeEnd, isCounter, false)
+	}
+	return floats[last].F
+}
+
+// interpolate performs linear interpolation between two points.
+// If isCounter is true and the second point's value is less than the first point's value,
+// it adds the first point's value to the second point's value to handle counter resets.
+// It then calculates the interpolated value at the given timestamp t.
+func interpolate(p1, p2 FPoint, t int64, isCounter, leftEdge bool) float64 {
+	y1 := p1.F
+	y2 := p2.F
+	if isCounter && y2 < y1 {
+		if leftEdge {
+			y1 = 0
+		} else {
+			y2 += y1
+		}
+	}
+
+	t1 := float64(p1.T)
+	t2 := float64(p2.T)
+	tf := float64(t)
+
+	return y1 + (y2-y1)*(tf-t1)/(t2-t1)
```

Comment:
isn't it slightly more accurate for precision to do the subtract first and then convert? presumably t1 is always < t2.
```suggestion
	return y1 + (y2-y1)*float64(t-p1.T)/float64(p2.T-p1.T)
```